aggregators/aggregator.go

```diff
@@ -603,23 +603,134 @@ func (a *Aggregator) processHarvest(
+		return harvestStats{}, fmt.Errorf("failed to process combined metrics ID %s: %w", cmk.ID, err)
+	}
+	return hs, nil
+}
+
+func (a *Aggregator) getAndLogOverflowStats(hs *harvestStats, cm *aggregationpb.CombinedMetrics, aggIvl time.Duration) {
+	hs.servicesOverflowed = hllSketchEstimate(cm.OverflowServicesEstimator)
+
+	logFunc := func(msg string, fields ...zap.Field) {
+		if a.cfg.OverflowLogging.Func != nil &&
+			(a.cfg.OverflowLogging.AggregationInterval == 0 || a.cfg.OverflowLogging.AggregationInterval == aggIvl) {
+			a.cfg.OverflowLogging.Func(msg, fields...)
+		}
+	}
+
+	// TODO(carsonip): Update this log message when global labels implementation changes
+	logFunc(fmt.Sprintf(""+
+		"Service limit reached, new metric documents will be grouped under a dedicated "+
+		"overflow bucket identified by service name '%s'. "+
+		"If you are sending global labels that are user-specific (e.g. client IP), it may cause "+
+		"high cardinality and lead to exhaustion of services.", overflowBucketName),
+		zap.Duration("aggregation_interval_ns", aggIvl),
+		zap.Int("limit", a.cfg.Limits.MaxServices),
+	)
+
+	// Flags to indicate global / "overall" limit reached per aggregation so that they are only logged once.
+	var txOverflowLogged, svcTxOverflowLogged, spanOverflowLogged bool
+
+	addOverflow := func(serviceName string, o *aggregationpb.Overflow, txGroups, svcTxGroups, spanGroups int) {
 		if o == nil {
 			return
 		}
-		hs.transactionsOverflowed += hllSketchEstimate(o.OverflowTransactionsEstimator)
-		hs.serviceTransactionsOverflowed += hllSketchEstimate(o.OverflowServiceTransactionsEstimator)
-		hs.spansOverflowed += hllSketchEstimate(o.OverflowSpansEstimator)
-	}
-	addOverflow(cm.OverflowServices)
-	for _, ksm := range cm.ServiceMetrics {
-		addOverflow(ksm.GetMetrics().GetOverflowGroups())
+		if overflowed := hllSketchEstimate(o.OverflowTransactionsEstimator); overflowed > 0 {
+			hs.transactionsOverflowed += overflowed
+			if serviceName != overflowBucketName {
+				if txGroups >= a.cfg.Limits.MaxTransactionGroupsPerService {
+					logFunc(fmt.Sprintf(""+
+						"Transaction group per service limit reached, "+
+						"new metric documents will be grouped under a dedicated bucket identified by transaction name '%s'. "+
+						"This is typically caused by ineffective transaction grouping, "+
+						"e.g. by creating many unique transaction names. "+
+						"If you are using an agent with 'use_path_as_transaction_name' enabled, it may cause "+
+						"high cardinality. If your agent supports the 'transaction_name_groups' option, setting "+
+						"that configuration option appropriately, may lead to better results.", overflowBucketName),
+						zap.String("service_name", serviceName),
+						zap.Duration("aggregation_interval_ns", aggIvl),
+						zap.Int("limit", a.cfg.Limits.MaxTransactionGroupsPerService),
+					)
+				} else if !txOverflowLogged {
+					logFunc(fmt.Sprintf(""+
+						"Overall transaction group limit reached, "+
```

Comment:
IIANM the only difference between this and the string above are the bits before the first comma. Rather than duplicating the entire string, can we use a common string (see other suggestion about package-level vars), and then add on that bit?